Gather etcd information from must-gather

palonsoro opened the pull request:
One of the most important things to check while getting information from a cluster is to check the cluster etcd health. This cannot be achieved by just getting some pod logs or kubernetes resources, but needs its own set of commands to be run in etcd pods.

In this pull request, I am adding a `gather_etcd` script to gather that information and I also made it to be run by default from main `gather` script.

Having this information on every must-gather would be of great help for support.

I'll be happy to help with any change or improvement required on this. And thanks in advance.

sferich888 commented on collection-scripts/gather:
```diff
@@ -43,3 +43,6 @@ done
 /usr/bin/gather_service_logs master
+
+# Gather etcd information
+/usr/bin/gather_etcd
```
Should this follow `gather_resources` (line 39)?

palonsoro replied:
I don't see a big difference so I can just move it up.

sferich888 commented on collection-scripts/gather_etcd:
```diff
@@ -0,0 +1,49 @@
+
+mkdir -p ${ETCD_LOG_PATH}
+
+echo "INFO: Starting getting etcd information"
```
Should this print the name of `RUNNING_ETCD_POD`?

palonsoro replied:
Good idea! I'll implement that.

wking replied:
This still doesn't include `RUNNING_ETCD_POD`.  Did you forget to commit and/or push?

palonsoro replied:
It was already added but in another log statement below. Please see: https://github.com/palonsoro/must-gather/blob/gather_etcd/collection-scripts/gather_etcd#L22

sferich888 replied:
Yea I still don't see where we are 'printing' the pod name. Can we add that here or to the items you reference.

palonsoro replied:
Hello, sorry the line number in my comment above is no longer correct due to changes happened after it.

Please look now at at: https://github.com/palonsoro/must-gather/blob/gather_etcd/collection-scripts/gather_etcd#L27

sferich888 commented on collection-scripts/gather_etcd:
```diff
@@ -0,0 +1,49 @@
+ocp4etcdctl alarm list > ${ETCD_LOG_PATH}/alarm_list.txt
+ocp4etcdctl alarm list -w json > ${ETCD_LOG_PATH}/alarm_list.json
+
+# List all etcd keys
```
This seems like it might be a security issue, in some 'rare' circumstances. 
Should we have a toggle for this? It also seems like a `costly` operation to preform. How much data is this going to pull back? @smarterclayton what is the averaged etcd count size shown from telemetry?

palonsoro replied:
My original intention with getting all the keys was to contrast whether users may claim something has been lost and/or some leftover from an unsupported version was present (due to a bug that prevented some api version migration, for example).

However, I did not thought about the security implications, so maybe I can just remove the line

sferich888 commented on collection-scripts/gather_etcd:
```diff
@@ -0,0 +1,49 @@
+
+echo "INFO: Starting getting etcd information"
+
+# member list
```
Should we run all of these in parallel?

palonsoro replied:
All those commands run very quickly (except for the check perf), so this is why I didn't make this parallel

sferich888 replied:
It's still ~8 `oc exec` calls. Depending on server load this could be cumbersome. Threading this lets all of this run at the same time and simply stop when completed, vs each call getting hung up and stalling the collection.

palonsoro replied:
Fair enough. I'll use the same parallelism scheme used in other gather scripts

palonsoro commented:
Changes implemented:
- Moved execution to the place requested
- Print etcd pod name as requested
- Removed key listing as suggested in order to prevent security issues
- `ETCDCTL_ENDPOINTS` computed from `member list` because the ones at the container may include bootstrap VM.

TO-DO:
- Parallelize execution

hexfusion commented on collection-scripts/gather_etcd:
```diff
@@ -0,0 +1,53 @@
+
+# Check perf
+echo "INFO: Performing etcdctl check perf"
+ocp4etcdctl check perf --load="s" > ${ETCD_LOG_PATH}/etcd_check_perf.txt
```
please do not do this, we monitor perf related metrics in Prometheus. If this command fails it will leave many keys in KVS. We are also capturing metrics through etcd events.

palonsoro replied:
Understood. I'll remove it.

hexfusion commented on collection-scripts/gather_etcd:
```diff
@@ -0,0 +1,53 @@
+BASE_COLLECTION_PATH="/must-gather"
+ETCD_LOG_PATH="${BASE_COLLECTION_PATH}/etcd_info"
+
+ETCD_CONTAINER='etcdctl'
```
can we clarify this ENV as `ETCDCTL_CONTAINER` so that I don't have a mini heart attack every time I see this code?

palonsoro replied:
Ok. No problem

palonsoro commented:
All requested changes implemented.

Thanks for all the feedback.

wking commented on collection-scripts/gather_etcd:
```diff
@@ -0,0 +1,61 @@
+#!/bin/bash
+
```
This probably needs guards like:

```sh
set -o nounset
set -o errexit
set -o pipefail
```

so that it doesn't blindly grind on if a step like calculating `RUNNING_ETCD_POD` fails.

palonsoro replied:
I am ok with `nounset` and `pipefail`. `errexit` would only be appropriate for `RUNNING_ETCD_POD` and `ETCDCTL_ENDPOINTS` calculation, because if one of the `ocp4etcdctl` commands return a non-zero exit code, the other ones should be tried.

It would be good to evaluate if any of the other files needs these guards, because none of them use them.

wking commented on collection-scripts/gather_etcd:
```diff
@@ -0,0 +1,61 @@
+RUNNING_ETCD_POD=$(oc get pod -n openshift-etcd --no-headers|grep Running| grep -o -m 1 '\S*etcd\S*')
+
+# We cannot rely on ETCDCTL_ENDPOINTS on container because it may contain bootstrap VM
+ETCDCTL_ENDPOINTS=$(oc exec ${RUNNING_ETCD_POD}  -n openshift-etcd -c ${ETCDCTL_CONTAINER} -- /bin/sh -c "etcdctl member list" | awk -F', ' '{printf "%s%s",sep,$5; sep=","}')
```
There are many missing quotes here, and through the rest of this script, that would fail ShellCheck:

```console
$ shellcheck collection-scripts/gather_etcd 

In collection-scripts/gather_etcd line 11:
ETCDCTL_ENDPOINTS=$(oc exec ${RUNNING_ETCD_POD}  -n openshift-etcd -c ${ETCDCTL_CONTAINER} -- /bin/sh -c "etcdctl member list" | awk -F', ' '{printf "%s%s",sep,$5; sep=","}')
                            ^-- SC2086: Double quote to prevent globbing and word splitting.


In collection-scripts/gather_etcd line 14:
    oc -n openshift-etcd exec -c ${ETCDCTL_CONTAINER} ${RUNNING_ETCD_POD} -- env ETCDCTL_ENDPOINTS=${ETCDCTL_ENDPOINTS} etcdctl $@
                                                      ^-- SC2086: Double quote to prevent globbing and word splitting.
                                                                                                   ^-- SC2086: Double quote to prevent globbing and word splitting.
                                                                                                                                ^-- SC2068: Double quote array expansions, otherwise they're like $* and break on spaces.


In collection-scripts/gather_etcd line 60:
wait ${PIDS[@]}
     ^-- SC2068: Double quote array expansions, otherwise they're like $* and break on spaces.

```

I think we should grow a ShellCheck job as a preflight for this repo.

palonsoro replied:
Let me fix that on `etcd_gather` script.

And I totally agree there should be an automated `ShellCheck` test, because none of the collection scripts completely passes it. It seems that there are quotes missing in many places. But I think I should just fix the etcd script at this PR, and fixing the other scripts should be another PR.

wking commented on collection-scripts/gather_etcd:
```diff
@@ -0,0 +1,61 @@
+
+# member list
+echo "INFO: Getting etcdctl member list"
+ocp4etcdctl member list -w table > ${ETCD_LOG_PATH}/member_list.txt &
```
Table output is convenient, but the JSON output should be sufficient, right?  I'd rather not bloat the must-gathers with redundant information, since folks can always write consuming code that pretty-prints the JSON if they want convenient inspection.

palonsoro replied:
Right: I pasted both JSON and table outputs because one of the most challenging aspects of dealing with must-gather is having all the information in machine-readable format only. However, I understand the need of no redundancy, so let me fix it.

palonsoro commented:
All the points outlined so far covered. Please let me know about any other feedback. And thanks everybody for the ideas.

sferich888 commented:
I have one minor nit to address but other than that it's. 
/lgtm 

@deads2k do we need a valid bug for this like tide says? Or is a Jira or something feature related acceptable?

palonsoro commented:
/retest
